06-building-decays.md

```diff
@@ -100,23 +100,60 @@ d0_sel = Selection("Sel_D0",
+from PhysSelPython.Wrappers import SimpleSelection
+d0_sel = SimpleSelection("Sel_D0",
+                         GenCombineParticles,
+                         [Pions, Kaons],
+                         DecayDescriptor='([D0 -> pi- K+]CC)',
+                         DaughtersCuts=d0_daughters,
+                         CombinationCut=d0_comb,
+                         MotherCut=d0_mother)
+```
+
+Note how we needed to import a different version of `CombineParticles` to make this work. This is because the LHCb algorithms are configured as singletons and it is mandatory to give them a name.
+
+> ## The LHCb singletons {.callout}
+> If we had tried to simply use `CombineParticles` inside our `SimpleSelection`, we would have seen it fail with the following error
+> 
+> ```output
+> NameError: Could not instantiate Selection because input Configurable CombineParticles has default name. This is too unsafe to be allowed.
+> ```
+>
+> The reason for this is that all LHCb algorithms need an explicit and unique name because they are *singletons*, and therefore an exception is thrown if we don't do that.
+> A [singleton](http://en.wikipedia.org/wiki/Singleton_pattern) is a software design pattern that restricts the instantiation of a class to one object. In our case, only one algorithm with a given name can be instantiated.
+> This allows to reuse and reload algorithms that have already been created in the configuration sequence, eg, we could have reloaded the `"Combine_D0"` `CombineParticles` by name and modified it (even in another file loaded in the same `gaudirun.py` call!):
+>
+> ```python
+> d0_copy = CombineParticles("Combine_D0")
+> print d0_copy.DecayDescriptor
+>```
+>
+> This is very useful to build complicated configuration chains.
+>
+> The solution for the `SimpleSelection` problem, in which we actually don't care about the `CombineParticles` name, is the `GaudiConfUtils.ConfigurableGenerators` package: it contains wrappers around algorithms such as `CombineParticles` or `FilterDesktop` allowing them to be instantiated without an explicit name argument.
+
 Now we can use another `CombineParticles` to build the D* with pions and the D0's as inputs, and applying a filtering only on the soft pion:
 
 ```python
 dstar_daughters = {'pi+': '(TRCHI2DOF < 3) & (PT > 100*MeV)'}
 dstar_comb = "(ADAMASS('D*(2010)+') < 400*MeV)"
 dstar_mother = "(abs(M-MAXTREE('D0'==ABSID,M)-145.42) < 10*MeV) & (VFASPF(VCHI2/VDOF)< 9)"
-dstar = CombineParticles('CombineDstar',
-                         DecayDescriptor='[D*(2010)+ -> D0 pi+]cc',
-                         DaughtersCuts=dstar_daughters,
-                         CombinationCut=dstar_comb,
-                         MotherCut=dstar_mother)
-dstar_sel = Selection('Sel_Dstar',
-                      Algorithm=dstar,
-                      RequiredSelections=[d0_sel, Pions])
+dstar_sel = SimpleSelection('Sel_Dstar',
+                            CombineParticles,
```

Comment:
It is. Damn copy-paste.

On Wed, Jun 3, 2015 at 10:11 AM, Tim Head <notifications@github.com> wrote:

> In 06-building-decays.md
> <https://github.com/lhcb/first-analysis-steps/pull/90#discussion_r31601710>
> :
>
> >  Now we can use another `CombineParticles` to build the D* with pions and the D0's as inputs, and applying a filtering only on the soft pion:
> >
> >  ```python
> >  dstar_daughters = {'pi+': '(TRCHI2DOF < 3) & (PT > 100*MeV)'}
> >  dstar_comb = "(ADAMASS('D*(2010)+') < 400*MeV)"
> >  dstar_mother = "(abs(M-MAXTREE('D0'==ABSID,M)-145.42) < 10*MeV) & (VFASPF(VCHI2/VDOF)< 9)"
> > -dstar = CombineParticles('CombineDstar',
> > -                         DecayDescriptor='[D*(2010)+ -> D0 pi+]cc',
> > -                         DaughtersCuts=dstar_daughters,
> > -                         CombinationCut=dstar_comb,
> > -                         MotherCut=dstar_mother)
> > -dstar_sel = Selection('Sel_Dstar',
> > -                      Algorithm=dstar,
> > -                      RequiredSelections=[d0_sel, Pions])
> > +dstar_sel = SimpleSelection('Sel_Dstar',
> > +                            CombineParticles,
>
> This should be GenCombineParticles ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lhcb/first-analysis-steps/pull/90/files#r31601710>.
>



-- 
Dr. Albert Puig Navarro
Laboratoire de Physique des Hautes Energies
Ecole Polytechnique Fédérale de Lausanne (EPFL)
BSP 614.4 (Cubotron UNIL) CH-1015 Lausanne
EPFL Phone: 021 6939808
CERN Phone: 72518